src/hardware/driver/dac.rs

```diff
@@ -0,0 +1,240 @@
+        pub const ENABLED: u32 = 1 << 11;
+    }
+    pub mod FSET {
+        pub const DISABLED: u32 = 0 << 10;
+        pub const ENABLED: u32 = 1 << 10;
+    }
+    pub mod VREFVAL {
+        pub const SPAN_5V_1_25V: u32 = 0010 << 6;
+        pub const SPAN_7_5V_1_25V: u32 = 0011 << 6;
+        pub const SPAN_10V_1_25V: u32 = 0100 << 6;
+        pub const SPAN_12_5V_1_25V: u32 = 0101 << 6;
+        pub const SPAN_15V_1_25V: u32 = 0110 << 6;
+        pub const SPAN_17_5V_1_25V: u32 = 0111 << 6;
+        pub const SPAN_20V_1_25V: u32 = 1000 << 6;
+        pub const SPAN_22_5V_1_25V: u32 = 1001 << 6;
+        pub const SPAN_25V_1_25V: u32 = 1010 << 6;
+        pub const SPAN_27_5V_1_25V: u32 = 1011 << 6;
+        pub const SPAN_30V_1_25V: u32 = 1100 << 6;
+    }
+    pub mod PDN {
+        pub const POWERUP: u32 = 0 << 4;
+        pub const POWERDOWN: u32 = 1 << 4;
+    }
+}
+
+#[allow(non_snake_case)]
+pub mod TRIGGER {
+    pub mod RCLTMP {
+        pub const RECAL: u32 = 1 << 8;
+    }
+    pub mod SRST {
+        pub const RESET: u32 = 1 << 6;
+    }
+    pub mod SCLR {
+        pub const CLEAR: u32 = 1 << 5;
+    }
+}
+
+#[allow(non_snake_case)]
+pub mod STATUS {
+    pub mod ALM {
+        pub const RECALIBRATED: u32 = 1 << 12;
+    }
+}
+
+pub enum Polarity {
+    AnodeGrounded,
+    CathodeGrounded,
+}
+
+#[allow(non_snake_case)]
+pub mod CONFIG2 {
+    pub mod DIS_TNH {
+        pub const DISABLED: u32 = 0 << 7;
+        pub const ENABLED: u32 = 1 << 7;
+    }
+    pub mod UP_RATE {
+        pub const UPDATE_RATE_1_MHZ_38_MHZ_SCLK: u32 = 000 << 4;
+        pub const UPDATE_RATE_0_9_MHZ_34_MHZ_SCLK: u32 = 001 << 4;
+        pub const UPDATE_RATE_0_8_MHZ_31_MHZ_SCLK: u32 = 010 << 4;
+        pub const UPDATE_RATE_1_2_MHZ_45_MHZ_SCLK: u32 = 011 << 4;
+        pub const UPDATE_RATE_0_5_MHZ_21_MHZ_SCLK: u32 = 100 << 4;
+        pub const UPDATE_RATE_0_45_MHZ_18_MHZ_SCLK: u32 = 101 << 4;
+        pub const UPDATE_RATE_0_4_MHZ_16_MHZ_SCLK: u32 = 110 << 4;
+        pub const UPDATE_RATE_0_6_MHZ_24_MHZ_SCLK: u32 = 111 << 4;
+    }
+}
+
+/// DAC value out of bounds error.
+#[derive(Debug)]
+pub enum Error {
+    Bounds,
+}
+
+/// A type representing a DAC sample.
+#[derive(Copy, Clone, Debug)]
+pub struct DacCode(u32);
+impl DacCode {
+    // DAC constants
+    const MAX_DAC_WORD: i32 = 1 << 20; // maximum DAC dataword (exclusive)
+    const VREF_DAC: f32 = 10.0; // Difference between positive and negaitiv reference pin
+    const R_OUT_LN: f32 = 40.0; // Low noise side output resistor
+    const R_OUT_HP: f32 = 0.68; // High power side output resistor
+}
+
+impl TryFrom<(f32, ChannelVariant)> for DacCode {
+    type Error = Error;
+    /// Convert an f32 representing a current int the corresponding DAC output code for the respective channel.
+    fn try_from(current_channel: (f32, ChannelVariant)) -> Result<Self, Error> {
+        let (current, channel) = current_channel;
+        let (r_out, is_inverted) = match channel {
```

Comment:
Just `-Self::R_OUT_LN` if inverted. And implement that as `fn transimpedance(&self) -> f32` on `ChannelVariant`.

And then your is masking and bounds check should just become check for overflow (recommend do it on paper first).

Reply:
What if I accidentally want to produce a negative current?

Reply:
Sorry, I cant think of a nice way to check for all the necessary bounds without explicit checks if we are inverted or not right now.. 
The current code is just there to illustrate my problem.

Reply:
https://github.com/quartiq/stabilizer/pull/616#pullrequestreview-1113981259

Reply:
But do review this. I may have missed something.

Reply:
I checked it for upper/lower bound in both cases. The bounds are as expected (and annoying since I immediately panic at 0.0 A for the inverted channels).